### ISSUE TITLE
An error occurred: DLL load failed while importing video_reader: The specified module could not be found.

### ISSUE THREAD

**Issue by NevermindNilas** [gcanat/video_reader-rs#11]
Hi,

```py
            import video_reader

            reader = video_reader.decode(input)
            for _ in range(totalFrames):
                frame = reader.read()
                if frame is None:
                    break

                frame = torch.from_numpy(frame)
                if isCudaAvailable:
                    frame = frame.to(device="cuda", non_blocking=True)
                readBuffer.put(frame)
```

Logs show:
```
An error occurred: DLL load failed while importing video_reader: The specified module could not be found.
```

**Comment by NevermindNilas** [comment]
My bad, I remembered there being a functionality for this but it turns out you are right.

I rerun the test in a new environment with the example code
![image](https://github.com/user-attachments/assets/756f4c09-b8ff-4013-b8c4-e4e75551a3eb)

```bash
Traceback (most recent call last):
  File "g:\import video_reader as vr.py", line 1, in <module>
    import video_reader as vr
  File "C:\Users\nilas\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\video_reader\__init__.py", line 1, in <module>
    from .video_reader import *
ImportError: DLL load failed while importing video_reader: The specified module could not be found.
```

This is the error that I am getting, it seems to happen on import. Could it be because I am using Python from MS?

**Comment by gcanat** [comment]
Yeah I'm not well versed into Windows platform, so I cannot really help. Did you install the pre-built package, ie with `pip install video_reader-rs` ? Maybe try to compile yourself by cloning the repo and doing `maturin develop -r`

**Comment by gcanat** [comment]
I got my hands on a Windows machine and was able to get it to work by doing the following:

Download ffmpeg full build: https://github.com/GyanD/codexffmpeg/releases/download/7.0/ffmpeg-7.1-full_build-shared.7z
Extract it somewhere, for example `C:\ffmpeg`

set FFMPEG_DIR env variable, for example in PowerShell: `$env:FFMPEG_DIR = "C:\ffmpeg\ffmpeg-7.0-full_build-shared"`

Install llvm if it is not already installed, for example: `choco install llvm`
Set LIBCLANG_PATH accordingly: `$env:LIBCLANG_PATH = "C:\Program Files\LLVM\bin"`

create a virtualenv: `conda create vr` and activate it `conda activate vr`

clone video_reader-rs repo and in the repo do: `maturin develop -r --features ffmpeg_6_0`

Now the last part is to add the dll directory in the python script before importing video_reader, Eg:
```python
import os
os.add_dll_directory("C:\\ffmpeg\ffmpeg-7.0-full_build-shared\\bin")
from video_reader import PyVideoReader
from time import perf_counter

filename = "/path/to/video.mp4"
chunk = 800

start = perf_counter()

vr = PyVideoReader(filename)
vid_len, height, width = vr.get_shape()

for start in range(0, vid_len, chunk):
    end = min(start+chunk, vid_len)
    frames = vr.decode(start_frame=start, end_frame=end)

duration = perf_counter() - start
print(f"FPS: {vid_len/duration:.2.f}")
```

Got around 1000 FPS on a 720x1280 videos with 4979 frames

**Comment by NevermindNilas** [comment]
```bash
(venv)  User    video_reader-rs   main ≡   3.12.7   1ms⠀   pip install maturin                                                              pwsh  20:
Collecting maturin
  Downloading maturin-1.7.4-py3-none-win_amd64.whl.metadata (18 kB)
Downloading maturin-1.7.4-py3-none-win_amd64.whl (7.4 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 35.4 MB/s eta 0:00:00
Installing collected packages: maturin
Successfully installed maturin-1.7.4

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip
(venv)  User    video_reader-rs   main ≡   3.12.7   2.874s⠀   $env:FFMPEG_DIR = C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmpeg-7.0-full_build-shared\bin
C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmpeg-7.0-full_build-shared\bin : The term
'C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmpeg-7.0-full_build-shared\bin' is not recognized as the name of a cmdlet, function, script file, or operable
program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:19
+ ... FMPEG_DIR = C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmp ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\User\D...uild-shared\bin:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

(venv)  User    video_reader-rs   main ≡   3.12.7   55ms⠀   $env:FFMPEG_DIR = "C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmpeg-7.0-full_build-shared\bin"
(venv)  User    video_reader-rs   main ≡   3.12.7   6ms⠀   $env:LIBCLANG_PATH = "C:\Program Files\LLVM\bin"                                 pwsh  20:
(venv)  User    video_reader-rs   main ≡   3.12.7   7ms⠀   maturin develop -r --features ffmpeg_6_0                                         pwsh  20:
    Updating crates.io index
     Locking 138 packages to latest compatible versions
      Adding hermit-abi v0.3.9 (latest: v0.4.0)
      Adding rustc-hash v1.1.0 (latest: v2.0.0)
      Adding video-rs v0.9.0 (latest: v0.10.1)
      Adding wasi v0.11.0+wasi-snapshot-preview1 (latest: v0.13.3+wasi-0.2.2)
      Adding windows-sys v0.52.0 (latest: v0.59.0)
      Adding yuvutils-rs v0.4.13 (latest: v0.5.0)
  Downloaded crossbeam-utils v0.8.20
  Downloaded icu_locid_transform v1.5.0
  Downloaded anstream v0.6.18
  Downloaded write16 v1.0.0
  Downloaded yoke-derive v0.7.4
  Downloaded zerofrom v0.1.4
  Downloaded icu_provider_macros v1.5.0
  Downloaded anstyle v1.0.10
  Downloaded synstructure v0.13.1
  Downloaded pyo3-macros v0.22.6
  Downloaded zerofrom-derive v0.1.4
  Downloaded writeable v0.5.5
  Downloaded crossbeam-deque v0.8.5
  Downloaded utf16_iter v1.0.5
  Downloaded zerovec-derive v0.10.3
  Downloaded icu_normalizer_data v1.5.0
  Downloaded rayon-core v1.12.1
  Downloaded serde_derive v1.0.214
  Downloaded url v2.5.3
  Downloaded zerovec v0.10.4
  Downloaded cc v1.1.37
  Downloaded idna v1.0.3
  Downloaded pyo3-ffi v0.22.6
  Downloaded numpy v0.22.1
  Downloaded rayon v1.10.0
  Downloaded icu_properties_data v1.5.0
  Downloaded icu_locid v1.5.0
  Downloaded icu_collections v1.5.0
  Downloaded serde v1.0.214
  Downloaded syn v2.0.87
  Downloaded icu_properties v1.5.1
  Downloaded icu_normalizer v1.5.0
  Downloaded pyo3-macros-backend v0.22.6
  Downloaded icu_provider v1.5.0
  Downloaded icu_locid_transform_data v1.5.0
  Downloaded crossbeam-epoch v0.9.18
  Downloaded yoke v0.7.4
  Downloaded pyo3-build-config v0.22.6
  Downloaded displaydoc v0.2.5
  Downloaded utf8_iter v1.0.4
  Downloaded tinystr v0.7.6
  Downloaded stable_deref_trait v1.2.0
  Downloaded pyo3 v0.22.6
  Downloaded litemap v0.7.3
  Downloaded idna_adapter v1.2.0
  Downloaded tokio v1.41.1
  Downloaded libc v0.2.162
  Downloaded 47 crates (4.5 MB) in 1.75s
📦 Including license file "F:\video_reader-rs\LICENSE"
🔗 Found pyo3 bindings
🐍 Found CPython 3.12 at F:\video_reader-rs\venv\Scripts\python.exe
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
   Compiling windows_x86_64_msvc v0.52.6
   Compiling autocfg v1.4.0
   Compiling target-lexicon v0.12.16
   Compiling stable_deref_trait v1.2.0
   Compiling once_cell v1.20.2
   Compiling glob v0.3.1
   Compiling libc v0.2.162
   Compiling regex-syntax v0.8.5
   Compiling crossbeam-utils v0.8.20
   Compiling minimal-lexical v0.2.1
   Compiling litemap v0.7.3
   Compiling writeable v0.5.5
   Compiling memchr v2.7.4
   Compiling bindgen v0.70.1
   Compiling either v1.13.0
   Compiling shlex v1.3.0
   Compiling smallvec v1.13.2
   Compiling bitflags v2.6.0
   Compiling icu_locid_transform_data v1.5.0
   Compiling rustc-hash v1.1.0
   Compiling num_cpus v1.16.0
   Compiling icu_properties_data v1.5.0
   Compiling itertools v0.13.0
   Compiling cc v1.1.37
   Compiling pkg-config v0.3.31
   Compiling vcpkg v0.2.15
   Compiling rayon-core v1.12.1
   Compiling cfg-if v1.0.0
   Compiling utf8_iter v1.0.4
   Compiling icu_normalizer_data v1.5.0
   Compiling write16 v1.0.0
   Compiling utf16_iter v1.0.5
   Compiling heck v0.5.0
   Compiling parking_lot_core v0.9.10
   Compiling rawpointer v0.2.1
   Compiling pin-project-lite v0.2.15
   Compiling scopeguard v1.2.0
   Compiling clang-sys v1.8.1
   Compiling utf8parse v0.2.2
   Compiling ffmpeg-next v7.1.0
   Compiling percent-encoding v2.3.1
   Compiling anstyle v1.0.10
   Compiling num-traits v0.2.19
   Compiling matrixmultiply v0.3.9
   Compiling memoffset v0.9.1
   Compiling lock_api v0.4.12
   Compiling nom v7.1.3
   Compiling aho-corasick v1.1.3
   Compiling form_urlencoded v1.2.1
   Compiling anstyle-parse v0.2.6
   Compiling tracing-core v0.1.32
   Compiling regex-automata v0.4.8
   Compiling unindent v0.2.3
   Compiling log v0.4.22
   Compiling indoc v2.0.5
   Compiling colorchoice v1.0.3
   Compiling is_terminal_polyfill v1.70.1
   Compiling cexpr v0.6.0
   Compiling windows-targets v0.52.6
   Compiling humantime v2.1.0
   Compiling bytes v1.8.0
   Compiling yuvutils-rs v0.4.13
   Compiling crossbeam-epoch v0.9.18
   Compiling quote v1.0.37
   Compiling regex v1.11.1
   Compiling pyo3-build-config v0.22.6
   Compiling num-integer v0.1.46
   Compiling num-complex v0.4.6
   Compiling windows-sys v0.59.0
   Compiling windows-sys v0.52.0
   Compiling libloading v0.8.5
   Compiling crossbeam-deque v0.8.5
   Compiling syn v2.0.87
   Compiling parking_lot v0.12.3
   Compiling anstyle-wincon v3.0.6
   Compiling anstyle-query v1.1.2
   Compiling rayon v1.10.0
   Compiling anstream v0.6.18
   Compiling socket2 v0.5.7
   Compiling mio v1.0.2
   Compiling pyo3-macros-backend v0.22.6
   Compiling pyo3-ffi v0.22.6
   Compiling pyo3 v0.22.6
   Compiling env_filter v0.1.2
   Compiling ndarray v0.16.1
   Compiling synstructure v0.13.1
   Compiling zerovec-derive v0.10.3
   Compiling displaydoc v0.2.5
   Compiling icu_provider_macros v1.5.0
   Compiling tracing-attributes v0.1.27
   Compiling tokio-macros v2.4.0
   Compiling env_logger v0.11.5
   Compiling zerofrom-derive v0.1.4
   Compiling yoke-derive v0.7.4
   Compiling tokio v1.41.1
   Compiling tracing v0.1.40
   Compiling zerofrom v0.1.4
   Compiling pyo3-macros v0.22.6
   Compiling yoke v0.7.4
   Compiling ffmpeg-sys-next v7.1.0
   Compiling zerovec v0.10.4
   Compiling tinystr v0.7.6
   Compiling icu_collections v1.5.0
error: failed to run custom build command for `ffmpeg-sys-next v7.1.0`

Caused by:
  process didn't exit successfully: `F:\video_reader-rs\target\release\build\ffmpeg-sys-next-cf6929ed34c7d4d3\build-script-build` (exit code: 101)
  --- stdout
  cargo:rustc-link-search=native=C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmpeg-7.0-full_build-shared\bin\lib
  cargo:rustc-link-lib=dylib=avcodec
  cargo:rustc-link-lib=dylib=avdevice
  cargo:rustc-link-lib=dylib=avfilter
  cargo:rustc-link-lib=dylib=avformat
  cargo:rustc-link-lib=dylib=avutil
  cargo:rustc-link-lib=dylib=swresample
  cargo:rustc-link-lib=dylib=swscale
  OPT_LEVEL = Some(3)
  OUT_DIR = Some(F:\video_reader-rs\target\release\build\ffmpeg-sys-next-82881c8c20434731\out)
  cargo:rerun-if-env-changed=VCINSTALLDIR
  VCINSTALLDIR = None
  cargo:rerun-if-env-changed=VSTEL_MSBuildProjectFullPath
  VSTEL_MSBuildProjectFullPath = None
  cargo:rerun-if-env-changed=VSCMD_ARG_VCVARS_SPECTRE
  VSCMD_ARG_VCVARS_SPECTRE = None
  cargo:rerun-if-env-changed=WindowsSdkDir
  WindowsSdkDir = None
  cargo:rerun-if-env-changed=WindowsSDKVersion
  WindowsSDKVersion = None
  cargo:rerun-if-env-changed=LIB
  LIB = None
  PATH = Some(F:\video_reader-rs\target\release\deps;F:\video_reader-rs\target\release;C:\Users\User\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib;F:\video_reader-rs\venv\Scripts;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\libnvvp;;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\NVIDIA Corporation\NVIDIA app\NvDLISR;C:\Program Files\dotnet\;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\7\;C:\ProgramData\chocolatey\bin;C:\Program Files\NVIDIA Corporation\Nsight Compute 2024.3.2\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\nodejs\;C:\Users\User\AppData\Local\Programs\Python\Python312\Scripts\;C:\Users\User\AppData\Local\Programs\Python\Python312\;C:\Users\User\.cargo\bin;C:\Users\User\AppData\Local\Microsoft\WindowsApps;C:\Users\User\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\User\AppData\Local\Programs\oh-my-posh\bin;C:\Users\User\AppData\Local\Microsoft\WinGet\Links;C:\Users\User\AppData\Roaming\npm)
  cargo:rerun-if-env-changed=INCLUDE
  INCLUDE = None
  HOST = Some(x86_64-pc-windows-msvc)
  cargo:rerun-if-env-changed=CC_x86_64-pc-windows-msvc
  CC_x86_64-pc-windows-msvc = None
  cargo:rerun-if-env-changed=CC_x86_64_pc_windows_msvc
  CC_x86_64_pc_windows_msvc = None
  cargo:rerun-if-env-changed=HOST_CC
  HOST_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some(cmpxchg16b,fxsr,sse,sse2,sse3)
  DEBUG = Some(false)
  cargo:rerun-if-env-changed=CFLAGS_x86_64-pc-windows-msvc
  CFLAGS_x86_64-pc-windows-msvc = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_pc_windows_msvc
  CFLAGS_x86_64_pc_windows_msvc = None
  cargo:rerun-if-env-changed=HOST_CFLAGS
  HOST_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  check.c
  check.c(3): fatal error C1083: Cannot open include file: 'libavutil/avutil.h': No such file or directory

  --- stderr
  cl : Command line warning D9035 : option 'o' has been deprecated and will be removed in a future release
  thread 'main' panicked at C:\Users\User\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ffmpeg-sys-next-7.1.0\build.rs:520:9:
  Compile failed
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit code: 101": `"cargo" "rustc" "--features" "ffmpeg_6_0" "--message-format" "json-render-diagnostics" "--manifest-path" "F:\\video_reader-rs\\Cargo.toml" "--release" "--lib"`
(venv)  User    video_reader-rs   main ≡  ?1   3.12.7   37.095s⠀                                                                           pwsh  20:
```

🤔 I can't replicate this. Seems like some .dll is missing on my end despite the choco llmv install with admin perms ( was necessary otherwise it seems to fail on my system ). This might be a me issue though.

I will try again later, fwiw, I use the official python install right now from `python.org` so no more MS bs.

Will try again at a later date.

**Comment by NevermindNilas** [comment]
Interesting, running pip install video-reader + the .dll load seems to at least technically work despite the maturin build failing, so I guess it's related specifically to that `ffmpeg bin` folder that's needed on the initial load.

Running a slightly altered code on a test video of shape, 1920x1080x3, fps 30, total frames roughly 18000 frames in total.
```py
import os
from time import time

os.add_dll_directory(r"C:\Users\User\Downloads\ffmpeg-7.0-full_build-shared\ffmpeg-7.0-full_build-shared\bin")
from video_reader import PyVideoReader

filename = r"F:\TheAnimeScripter\input\10minTest.mp4"
chunk = 800

start_time = time()

vr = PyVideoReader(filename)
vid_len, height, width = vr.get_shape()

for start in range(0, vid_len, chunk):
    end = min(start + chunk, vid_len)
    frames = vr.decode(start_frame=start, end_frame=end)

end_time = time()
duration = end_time - start_time
print(f"FPS: {vid_len / duration}")

```

```bash
╰─ & f:/video_reader-rs/venv/Scripts/python.exe "f:/video_reader-rs/import os.py"
FPS: 425.6823250843223
```

I've also ran it on a 1280x720, with a total of 4500frames
```bash
╰─ & f:/video_reader-rs/venv/Scripts/python.exe "f:/video_reader-rs/import os.py"
FPS: 888.3031078685938
```

My CPU is a 13700k ( with a slight curve OC ).
